k8scrdclient/k8scrdclient.go

```diff
@@ -120,23 +120,34 @@ func (c *CRDClient) ensureCreated(ctx context.Context, crd *apiextensionsv1beta1
 	return nil
 }
 
-// ensureStatusSubresourceCreated ensures if the CRD has a status subresource
-// it is created. This is needed if a previous version of the CRD without the
-// status subresource is present.
-func (c *CRDClient) ensureStatusSubresourceCreated(ctx context.Context, crd *apiextensionsv1beta1.CustomResourceDefinition, b backoff.Interface) error {
-	if crd.Spec.Subresources == nil || crd.Spec.Subresources.Status == nil {
-		// Nothing to do.
-		return nil
-	}
-
+// ensureUpdated ensures if the CRD changed it is updated accordingly. This is
+// needed if e.g. a previous version of the CRD without the status subresource
+// is present where it should actually be set. Another example would be the CRD
+// apiversion changing, which tends to happen every now and then over the
+// runtime object lifecycle and community adoption.
+func (c *CRDClient) ensureUpdated(ctx context.Context, crd *apiextensionsv1beta1.CustomResourceDefinition, b backoff.Interface) error {
 	o := func() error {
-		manifest, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		latest, err := c.k8sExtClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
 		if err != nil {
 			return microerror.Mask(err)
 		}
 
-		if manifest.Spec.Subresources == nil || manifest.Spec.Subresources.Status == nil {
-			crd.SetResourceVersion(manifest.ResourceVersion)
+		var changed bool
+		{
+			if !crdAPIVersionEqual(crd, latest) {
+				changed = true
+			}
+			if !crdStatusEqual(crd, latest) {
+				changed = true
+			}
+			if !crdValidationEqual(crd, latest) {
+				changed = true
+			}
+		}
+
+		if changed {
+			crd.SetResourceVersion(latest.ResourceVersion)
```

Comment:
This approach drops changes made to CR between the time when `crd` was read and `latest` was updated.

1. Read `crd`.
2. Some other party updates the object.
3. Execution reaches the following `Update()`.

I'm not sure what would CRD except us from here, but I'd rather write the logic in following way:
1. Read CR
2. Apply changes to CR
3. `Update(cr)`
4. If `Update()` failed on conflict, go to `1.`.

That way changes don't get lost.

If there's an edge case and some other program reconciled the object for some specific changes, this could even lead to update loop between two parties.

Reply:
I don't see the problem. We read, check and update in a backoff operation. On conflict we start over again. `crd` is the desired state.

Reply:
Problem is this:

- A reads `crd`
- B reads `crd`
- B modifies `crd.Status`
- A modifies `crd.Spec`
- B writes `crd`
- B's write succeeds
- A writes `crd`
- A's write fail because its `ResourceVersion` is outdated
- A reads `latest`
- A copies `latest.ResourceVersion` into `crd.ResourceVersion`
- A writes `crd`
- A's write succeeds
- B observes that `crd.Status` modifications are gone.

Reply:
We had a chat about this. The problem we have is having CRDs in different versions, which especially becomes an issue with multiple operators running as we intend to do with VOO. We need to figure out how to do CRD versioning.

Reply:
This is a problem in general but the code here shouldn't override anything. Update should fail with conflict if the CRD was modified in between because the resourceVersion changed.

Reply:
Let's please not continue with a discussion about the problems in the comments. Let's have a call when there are thoughts to be exchanged.

Reply:
I actually read @tuommaki comments more carefully and I 100% agree with him. I also don't have immediate solution to the problem. I also see you had a chat. So I guess the call isn't necessary? Unless you want to brainstorm something. Then ping me on Slack.

Reply:
See https://gigantic.slack.com/archives/C2MS2VB37/p1579862842041600.